Suggestions and changes to "Commands" and "Events" architecture sections (#248)
boostercloud/booster: docs/README.md

![architecture](./img/booster-arch.png)

The public interface of a Booster application is just `Commands` and `ReadModels`. Booster proposes an entirely different approach to the Model-View-\* and CRUD frameworks. With Booster, clients submit commands, query the read models, or subscribe to them for receiving real-time updates thanks to the out of the box [GraphQL API](#graphql-api)

Booster applications are event-driven and event-sourced so, **the source of truth is the whole history of events**. When a client submits a command, the `CommandHandler` _wakes up_ and executes its logic. Optionally, it can _register_ as many `Events` as needed. The framework caches the current state by automatically _reducing_ all the registered events into `Entities`. Interested parties can _react_ to events via `EventHandlers`, and finally, the _projection_ functions transform the entities into `ReadModels`.

In this chapter you'll walk through these concepts and its details.

### 1. Command and command handlers

> Booster is different than MVC frameworks in which you typically implement controller classes with the four CRUD methods per model. Instead of that, you define commands, which are the user actions when interacting with an application. This approach fits very well with Domain-Driven Design. Depending on your application's domain, some examples of commands would be: `RemoveItemFromCart`, `RatePhoto`, `AddCommentToPost`, etc. Although, you can still have `Create*`, `Delete*`, or `Update*` commands when they make sense.

Author: Not sure why this is marked as changed 😳

Author: Oh! I changed "*five* CRUD methods..." by "*four* CRUD methods..."

Curious: Does anybody know why it was written "five"? @boostercloud/booster-core @itrion

> Booster is different than MVC frameworks in which you typically implement controller classes with the four CRUD methods per model. Instead of that, you define commands, which are the user actions when interacting with an application. This approach fits very well with Domain-Driven Design. Depending on your application's domain, some examples of commands would be: `RemoveItemFromCart`, `RatePhoto`, `AddCommentToPost`, etc. Although, you can still have `Create*`, `Delete*`, or `Update*` commands when they make sense.

Reviewer: ```suggestion
Booster is different than MVC frameworks in which you typically implement controller classes with CRUD methods per model. Instead of that, you define commands, which are the user actions when interacting with an application. This approach fits very well with Domain-Driven Design. Depending on your application's domain, some examples of commands would be: `RemoveItemFromCart`, `RatePhoto`, `AddCommentToPost`, etc. Although, you can still have `Create*`, `Delete*`, or `Update*` commands when they make sense.
```

Just curious about this, doesn't "CRUD" includes specifically 4 methods?

Reviewer 2: Yes, but Read is done through the Query API

Author: Done b40504a

> There is an architectural splitting between commands and command handlers though they *live* under the same file. The command is the class with the `@Command` decorator, and the generated method called `handle` is the command handler. That is because Booster adopts several concepts from functional programming; the separation between data structures and data transformations is one of them. In Booster, a command looks like this:

Author: Yeah, at least to be consistent with the other concepts

}
```

Every time you submit a command through the GraphQL API, Booster calls the command handler function for the given command. Commands are part of the public API, so you can define authorization policies for them. They are also the place for validating input data before registering events into the event store because they are immutable once there.

#### Commands naming convention

> Semantics is very important in Booster as it will play an essential role in designing a coherent system. Your application should reflect your domain concepts, and commands are not an exception. Although you can name commands in any way you want, we strongly recommend you to name them starting with verbs in imperative plus the object being affected. If we were designing an e-commerce application, some commands would be:

Author: Apparently the noun goes always in the plural form

boost new:command CreateProduct --fields sku:SKU displayName:string description:string price:Money
```

> The above generator will automatically create a file called `CreateProduct.ts` with a TypeScript class of the same name under the `commands` directory. You can still create (or modify) the command manually. Since the generator is not doing any *magic*, all you need is a class decorated as `@Command`. Anyway, we recommend you always to use the generator, because it handles the boilerplate code for you.

Reviewer 3: @alvaroloes, something I've learned is to avoid explicit mentions to the position of the tables, snippets, graphics, etc. because if you move them, you don't have to change the texts with the reference.

```suggestion
The generator will automatically create a file called `CreateProduct.ts` with a TypeScript class of the same name under the `commands` directory. You can still create (or modify) the command manually. Since the generator is not doing any *magic*, all you need is a class decorated as `@Command`. Anyway, we recommend you always to use the generator, because it handles the boilerplate code for you.
```

I learned it using LaTex because it is in charge of placing the visual elements wherever they fit better, so you actually don't know where they'll be after the build. I know we are not using LaTex here but I still consider it a very good tip as we could be placing snippets in a different position after building the final docs

Author: Amazing tip Johan. 100% agree

Note:
> Generating a command with the same name as an already existing one will override its content. Soon, we will display a warning before overwriting anything.

#### The command handler function

> Each command class must have a method called `handle`. This function is the command handler, and it will be called by the framework every time one instance of this command is submitted. Inside it you can run validations, return errors, query entities to make decisions and register relevant domain events.

Reviewer 4: ```suggestion
Each command class must have a method called `handle`. This function is the command handler, and it will be called by the framework every time one instance of this command is submitted. Inside the handler you can run validations, return errors, query entities to make decisions and register relevant domain events.
```

> Each command class must have a method called `handle`. This function is the command handler, and it will be called by the framework every time one instance of this command is submitted. Inside it you can run validations, return errors, query entities to make decisions and register relevant domain events.

Reviewer: ```suggestion
Each command class must have a method called `handle`. This function is the command handler, and it will be called by the framework every time one instance of this command is submitted. From command handlers, you can run validations, return errors, query entities to make decisions, and register relevant domain events.
```

Author: Applied already a similar suggestion above. Thanks!

Author: Included the extra comma: b40504a

## Features

> ### Authentication and Authorization

Reviewer 4: This might require a regeneration of the table of content

Author: done! b40504a
